File: CS50x/psets/pset06/sentimental-credit/credit.py
import math


def main():
    card_number = input("Number: ")  # Read the card number as a string
    card_type(card_number)  # Print wich company the card is from


def luhn_algorithm(card_number):  # Sum of the numbers in the card using luhn_algorithm
    sum = 0
    for i in range(len(card_number)):
        digit = int(card_number[len(card_number) - 1 - i])  # Begins at the last digit
        if i % 2 == 0:
            sum += digit
        else:
            digit *= 2
            sum += digit // 10 + digit % 10
    return sum


def card_type(card_number):
    size = len(card_number)  # The number of digits the card has
    sum = luhn_algorithm(card_number)

    # Important information on the card
    # print(f"Size: {size}")
    # print(f"Sum: {sum}")
    # print("Digits: " + card_number[0:2])

    if sum % 10 == 0:
        if card_number[0:2] in ["34", "37"] and size == 15:
            print("AMEX")
        elif card_number[0:2] in ["51", "55"] and size == 16:
            print("MASTERCARD")
        elif card_number[0] == "4" and 13 <= size <= 16:
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()

File: pset01/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long long CardNum;
    do
    {
        CardNum = get_long_long("Number: ");
    }
    while (CardNum < 0);

    int size = 0;
    {
        long long x = 1;
        while (x < CardNum)
        {
            x = x * 10;
            size++;
        }
    }

    long long div = 10;
    int sum = 0, order = size, t, t1, t2;
    for (int i = 0; i < size; i++, order--, div *= 10)
    {
        printf("Número #%i: %lli\n", (order), ((CardNum % div) / (div / 10)));

        if (i % 2 == 0)
        {
            sum += ((CardNum % div) / (div / 10));
        }
        else
        {
            sum += ((((CardNum % div) / (div / 10)) * 2) / 10) + (((CardNum % div) / (div / 10)) * 2) % 10;
        }

        if (order == 1)
        {
            t1 = ((CardNum % div) / (div / 10));
        }
        if (order == 2)
        {
            t2 = ((CardNum % div) / (div / 10));
        }
    }

    t = (t1 * 10) + t2;
    printf("First two digits: %i\n", t);
    printf("Size: %i\n", size);
    printf("Sum: %i\n", sum);

    if (sum % 10 == 0)
    {
        if ((t == 34 || t == 37) && size == 15)
        {
            printf("AMEX\n");
        }
        else if ((t > 50 && t < 56) && size == 16)
        {
            printf("MASTERCARD\n");
        }
        else if (t1 == 4 && size > 12 && size < 17)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}